feat!: entrypoint and fee payment simplification

Thunkar opened the pull request:
# Context for the reviewer

This PR initially attempted to simplify the way fee payment methods were forwarded (if at all!) to the `Wallet` software. Historically, we have allowed these fee payment methods to be assembled "app side" (even if we didn't have a clear separation between wallets and apps). After the wallet refactor and new interface, it became clear this is not  a good design since ultimately wallets (read: users) should decide how the want to pay fees. Unfortunately this is not as simple as that, because: 

- Thanks to fee abstraction, fee payment methods are *just contract function calls*. Realistically, there's nothing preventing an app from crafting a transaction that pays for its own fee and sending it to the wallet. We even had a case in our repo: Account contract deployments
- There are legitimate use cases for apps assembling their own fee-paying transactions, such as sponsoring, subscriptions, etc.

With this in mind, now please bear with me

# Entrypoints and fee payment methods

Up until now, entrypoints of account contracts had knowledge of which calls were related to fee payment and which ones were "app" calls. Even if this might seem practical, for the aforementioned reasons it didn't make much sense. Nothing prevented an app from ignoring this convention, and to add insult to injury the account contract was the one responsible of ending the setup (non-revertible) phase of the transaction, which would break apps that tried to put fee calls in the "app" part.

What's even worse, ending the setup of the tx on the account entrypoint lead to the very unfortunate consequence of FPCs never having the assurance of them being executed on that phase. A malicious account contract could just put the calls to pay fees on the revertible side of the tx and make them fail on purpose, while keeping purely private app logic in the non-revertible phase...essentially getting the side effects for free.

# What's changed

- Account contract entrypoints do not distinguish between app and fee calls now. This reflects the fact that these calls are indeed indistinguishable and we should not attempt to add a convention that is broken from the get-go.
- Ending the setup phase is the responsibility of *whomever wants to get paid*. As of now, ending it is the only way contracts can get assurance on whether their side effects are revertible or not.
- Fee payment methods *can* be used on the app side, and we provide some help with that:
    - Interactions (`ContractFunctionInteraction`, `DeployMethod`, etc) allow setting a  `fee.paymentMethod` option just like before. HOWEVER, what this will do is obtain the `ExecutionPayload` for that particular fee payment method and it will merge it with whatever the interaction was doing. This works like this because *it is the only sane way*. If the app wants to "bundle" a fee payment method, it needs intimate knowledge of how the calls are assembled. A great example of this is account contract self-deployments: the fee payment method has to come AFTER the construction of the contract, or it won't be able to validate an authwit (the public key not does not exist yet) and it won't be able to pay for its own fee. This only used to work because we sneakily inverted the order of execution for the `MulticallEntrypoint` to allow for this use case. Terrible.
    - `FeeJuicePaymentMethod` CANNOT be bundled into an interaction. For two reasons: first of all, it's very weird that an app tells the wallet: "hey, here's your tx with a fee payment method, do not a attempt to set one. By the way, the account I'm sending this for is paying". On the other hand, the app would have to convey to the account that they're the fee payer and ALSO that they have to end the setup phase. Conceptually, paying the fee with your own account contract is a *feature* of the account contract itself (which it might disallow for privacy reasons!) so the only one that could possible set it reliably is the wallet.

# Happy consequences

- Account entrypoints can now do 5 functions by default, instead of 4 + to optional fee ones. The entrypoint is also cheaper!
- Fee payment methods can still be bundled on the "app side"
- The `Wallet` interface is now 100% serializable
- FPCs now can be sure that they executed on the setup phase

# Not so happy consequences

- Not being able to say `paymentMethod: FeeJuice` feels weird. Once you think about it, it makes sense (from the app side). The wallet of course will still be able to do that (if its account contracts support it)
- The FeeJuice contract now has 2 separate `claim` methods, one that can be invoked anytime, and a specific "fee payment" one that also ends the setup phase.
- The CLI wallet is in a weird place rn and `cancel_tx` command is broken. Will address it in a followup PR

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -83,15 +84,21 @@ export class EmbeddedWallet extends BaseWallet {
-      new SponsoredFeePaymentMethod(sponsoredFPCContract.instance.address);
+    const paymentMethod = !userFeeOptions?.embeddedPaymentMethodFeePayer
+      ? new SponsoredFeePaymentMethod(sponsoredFPCContract.instance.address)
+      : undefined;
```
```suggestion
    const paymentMethod = userFeeOptions?.embeddedPaymentMethodFeePayer
      ? undefined : new SponsoredFeePaymentMethod(sponsoredFPCContract.instance.address);
```
Think it's more readable when not having to do the bool flip.

Could you add here a comment why we populate the payment method only when embeddedPaymentMethodFeePayer is undefined and why the sponsored fee payment method is used?

Thunkar replied:
This has evolved, but still I prefer the actual "something" return value first, I find it easier to read. Different brain damage.

benesjan commented on yarn-project/entrypoints/src/interfaces.ts:
```diff
@@ -19,6 +17,17 @@ export type TxExecutionOptions = {
   txNonce?: Fr;
+  /**
+   * Account contracts usually support paying transaction fees using their own fee juice balance, which is configured
+   * setting this flag to true.
```
```suggestion
   * by setting this flag to true.
```

benesjan commented on noir-projects/aztec-nr/aztec/src/authwit/account.nr:
```diff
@@ -28,27 +26,30 @@ impl<Context> AccountActions<Context> {
+    pub fn entrypoint(
+        self,
+        app_payload: AppPayload,
+        is_fee_payer: bool,
+        end_setup: bool,
+        cancellable: bool,
```
These args are now undocumented. Would describe here mainly the motivation of why we would set each of these flags and its impact (and that end setup can be called only once and it's impact on FPCs).

Aztec.nr is more stable than Aztec.js so this seems like a good place to have such explanation.

Thunkar replied:
Agreed, added comments. I removed the stale ones but didn't realize the new ones weren't in

benesjan commented on noir-projects/aztec-nr/aztec/src/authwit/account.nr:
```diff
@@ -28,27 +26,30 @@ impl<Context> AccountActions<Context> {
+        if is_fee_payer {
+            self.context.set_as_fee_payer();
+        }
+        if end_setup {
```
If `is_fee_payer` is set to true then we would always want `end_setup` to be true as well and vice versa, right?

If yes, then it makes sense to merge these flags.

nventuro replied:
Yea I'm also confused by the possibility of setting the fee payer but not ending setup. The PR description seems to hint at the opposite being true - FPCs always end setup (to ensure they're in setup).

Thunkar replied:
See this: https://github.com/AztecProtocol/aztec-packages/pull/17375#discussion_r2392352107

benesjan commented on docs/versioned_docs/version-v3.0.0-nightly.20250930/developers/docs/concepts/call_types.md:
```diff
@@ -103,11 +103,11 @@ Contract functions marked with `#[private]` can only be called privately, and as
 
 Private functions from other contracts can be called either regularly or statically by using the `.call()` and `.static_call` functions. They will also be 'executed' (i.e. proved) in the user's device, and `static_call` will fail if any state changes are attempted (like the EVM's `STATICCALL`).
```
was there some real change here or is it just fmt?

Thunkar replied:
fmt

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -301,9 +315,25 @@ export class EmbeddedWallet extends BaseWallet {
+      await feeOptions.paymentMethod?.getExecutionPayload();
+    const maybeFeePayer =
+      (await feeOptions.paymentMethod?.getFeePayer()) ??
+      feeOptions.embeddedPaymentMethodFeePayer;
```
This is quite convoluted. Would put here a comment when this is undefined and what happens if it's undefined.

Thunkar replied:
It is. That's why my latest commits changed it 😛

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -301,9 +315,25 @@ export class EmbeddedWallet extends BaseWallet {
+      feeExecutionPayload ?? ExecutionPayload.empty(),
       executionPayload,
-      feeOptions,
+    ]);
```
Merging execution payload with empty payload seems strange. Think it would be clearer if we just didn't call it when there is no fee execution payload.

nventuro replied:
Agreed, it makes `merge` more complicated. I'd do e.g.

```ts
const finalExecutionPayload = feeExecutionPayload ?  
    mergeExecutionPayloads([feeExecutionPayload, executionPayload]) :
    executionPayload;
```

nventuro replied:
I know this was not added in this PR, but this shows up _so_ often all over the place that it's becoming a bit problematic. It's not even trivial to figure out that merging with an empty exec payload is ok. We can do this later, but i'd seriously consider removing the empty exec payload function, and possibly having a `mergeWithFee` or w/e that takes an optional fee exec payload. It's adding mental overhead and is brittle.

benesjan commented on noir-projects/noir-contracts/contracts/fees/fpc_contract/src/main.nr:
```diff
@@ -101,6 +101,8 @@ pub contract FPC {
 
         // Set the FPC as the fee payer of the tx.
         context.set_as_fee_payer();
+        // End the setup phase, from now on all side effects are revertible
+        context.end_setup();
```
Unrelated to your PR but it's very Interesting that this is now being called in the same place and it seems to be the case all around the codebase.

Makes me think whether this should be just one function 🤔

WDYT?

nventuro replied:
Was thinking the same

Thunkar replied:
Unfortunately not, and it makes me very sad.

An FPC is usually the fee payer AND it always calls `end_setup` to ensure it's getting paid (the logic runs in the non-revertible phase). However, there is one exception:

The infamous `FeeJuiceWithClaim` payment method is divided into 2. The account contract sets itself as fee payer, but the actual end of the setup phase has to be invoked by the `FeeJuice` contract AFTER the L1toL2 message is retrieved and the public balance increase enqueued. Sad.

benesjan commented on noir-projects/noir-contracts/contracts/protocol/fee_juice_contract/src/main.nr:
```diff
@@ -35,9 +42,23 @@ pub contract FeeJuice {
+        FeeJuice::at(context.this_address())._increase_public_balance(to, amount).enqueue(context);
+    }
+
+    #[private]
+    fn claim(to: AztecAddress, amount: u128, secret: Field, message_leaf_index: Field) {
+        claim_helper(&mut context, to, amount, secret, message_leaf_index);
+    }
```
When would we want to call this instead of the end setup one below?

nventuro replied:
This would be when claiming but unrelated to FPCs etc - just a claim. I agree that it'd be good to have a descrption explaining why both exist and use cases for each.

Thunkar replied:
You can call `claim` as a standalone operation, if you just want to increase your fee juice balance. That interaction would not be whitelisted in the node, since it doesn't have to.

benesjan commented on playground/src/components/home/components/Landing.tsx:
```diff
@@ -371,26 +371,24 @@ export function Landing() {
-          paymentMethod: await accountManager.getSelfPaymentMethod(feePaymentMethod),
+          paymentMethod,
         },
-        universalDeploy: true,
```
Why we no longer need universalDeploy?

Thunkar replied:
All account contracts are deployed as `universalDeploy: true`, meaning this was the only correct configuration. The new `DeployAccountMethod` takes care of setting the default, and it is explained there: 
<img width="755" height="143" alt="Screenshot 2025-09-30 at 19 28 08" src="https://github.com/user-attachments/assets/27a7b7aa-ea66-477e-bb61-11a396d16c28" />

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
   : any;
+
+/**
+ * Transforms and cleans up the higheer level SendMethodOptions defined by the interaction into
```
```suggestion
 * Transforms and cleans up the higher level SendMethodOptions defined by the interaction into
```

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
+}
+
+/**
+ * Transforms and cleans up the higheer level SimulateMethodOptions defined by the interaction into
```
```suggestion
 * Transforms and cleans up the higher level SimulateMethodOptions defined by the interaction into
```

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
+      // try to reuse as much as possible while still allowing
+      // manual override. CAREFUL: this can cause mismatches during proving
+      gasSettings: {
+        ...options?.fee?.paymentMethod?.getGasSettings(),
```
```suggestion
        ...options.fee?.paymentMethod?.getGasSettings(),
```

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
+      // manual override. CAREFUL: this can cause mismatches during proving
+      gasSettings: {
+        ...options?.fee?.paymentMethod?.getGasSettings(),
+        ...options?.fee?.gasSettings,
```
```suggestion
        ...options.fee?.gasSettings,
```

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
+      // try to reuse as much as possible while still allowing
+      // manual override. CAREFUL: this can cause mismatches during proving
+      gasSettings: {
+        ...options?.fee?.paymentMethod?.getGasSettings(),
```
```suggestion
        ...options.fee?.paymentMethod?.getGasSettings(),
```

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
+      // manual override. CAREFUL: this can cause mismatches during proving
+      gasSettings: {
+        ...options?.fee?.paymentMethod?.getGasSettings(),
+        ...options?.fee?.gasSettings,
```
```suggestion
        ...options.fee?.gasSettings,
```

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
+  return {
+    ...options,
+    fee: {
+      embeddedPaymentMethodFeePayer: await options.fee?.paymentMethod?.getFeePayer(),
```
I don't really know what to do with this and whether it even makes sense to think about given that all fees might change but these types just seem so complicated.

Like I am absolutely not able to figure out when `embeddedPaymentMethodFeePayer` would be populated because `fee` on option might be undefined and paymentMethod on the fee might be undefined but if these 2 are defined I get thee feepayer from it. There are just too many dimensions of undefined-ness.

Might be a good use of time to draw some diagrams at the onsite and figure out if all these types could be streamlined somehow.

Thunkar replied:
I think this is pretty straightforward as long as you understand the following:

Fee payment methods are *usually* added by the wallet, and the app doesn't have to care. HOWEVER, there are certain situations where an app can embed a fee payment method in the interaction, for example to sponsor the user. The undefined-ness comes because all levels of `opts.fee.paymentMethod` are optional. 
In case an app embeds a payment method, we need to provide a hint to the wallet so it doesn't add a new fee payment method (since by definition, they are indistinguishable from app calls). This is explained on the wallet side: 
<img width="757" height="201" alt="Screenshot 2025-09-30 at 19 26 39" src="https://github.com/user-attachments/assets/ee4dd48d-09bb-435a-bb62-67276630a097" />



Bonus points:
Why the address and not a boolean? Well, in case the fee payment method is `FeeJuiceWithClaim`, the account contract has to set itself as fee payer...meaning a boolean is not enough.

Anyhow...yeah, we should have an onsite session on wallets, fee payment methods and how apps can set them

nventuro replied:
I searched for this function because the usage seemed concerning. Some comments:

- the distinction between SendOptions and SendMethodOptions is extremely not clear to me
- 'sanitize' is a bit of an odd choice of word
- I agree with Jan's intuition re. it being hard to thing about the variants. Again I go back to maybe needing better types to represent that possibilities? At the very least we need to look at all options and streamline the naming, e.g. it's not obviously clear what 'embedded' means

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -72,3 +113,57 @@ export type SimulationReturn<T extends boolean | undefined> = T extends true
+}
+
+/**
+ * Transforms and cleans up the higheer level ProfileMethodOptions defined by the interaction into
```
```suggestion
 * Transforms and cleans up the higher level ProfileMethodOptions defined by the interaction into
```

benesjan commented on yarn-project/aztec.js/src/fee/fee_payment_method.ts:
```diff
@@ -0,0 +1,36 @@
+   * The gas settings (if any) used to compute the
+   * execution payload of the payment method
+   */
+  getGasSettings(): GasSettings | undefined;
```
Why are the gas settings here defined only sometimes?

Is it the case that we fallback to the defaults when the fee payment method doesn't define it?

And if yes wouldn't it be better to just force the fee payment method to just get the defaults on its own? Might make it all easier to navigate.

My general feeling is that too many `undefined`s make  the types less useful as then they stop giving a guarantee of the ting being correct.

Thunkar replied:
Getting the default gassettings is complicated since it involves a call to the node + setting a padding, which is something only known to:

- The wallet (if it's the one adding the fee payment method because the interaction didn't include it)
- The app (if it's embedding the fee payment method in the interaction)

This is probably the part I like the less about the refactor, but it's just the way it is.

benesjan commented on yarn-project/aztec.js/src/contract/batch_call.ts:
```diff
@@ -3,13 +3,17 @@ import { type FunctionCall, FunctionType, decodeFromAbi } from '@aztec/stdlib/ab
   constructor(
     wallet: Wallet,
-    protected calls: BaseContractInteraction[],
+    protected interactions: (ExecutionPayload | BaseContractInteraction)[],
```
Why did this PR forced you to make the type here more complicated?

Thunkar replied:
Nothing, but it is more useful this way. I can revert since it ended up being unused here.

benesjan commented on yarn-project/aztec.js/src/contract/batch_call.ts:
```diff
@@ -29,13 +33,10 @@ export class BatchCall extends BaseContractInteraction {
+    const feeExecutionPayload = options.fee?.paymentMethod
+      ? await options.fee.paymentMethod.getExecutionPayload()
+      : ExecutionPayload.empty();
+    return mergeExecutionPayloads([feeExecutionPayload, ...requests]);
```
Also here would call the merge only if feeExecutionPayload is real.

Thunkar replied:
I find the single control flow (not really, but looks like it) option better

benesjan commented on yarn-project/aztec.js/src/contract/batch_call.ts:
```diff
@@ -113,7 +114,14 @@ export class BatchCall extends BaseContractInteraction {
+        }
+        return i;
+      }),
+    );
```
This change is sad.

benesjan commented on yarn-project/aztec.js/src/utils/authwit.ts:
```diff
@@ -269,16 +266,13 @@ export class SetPublicAuthwitContractInteraction extends ContractFunctionInterac
   public override simulate<T extends SimulateMethodOptions>(
-    options: T,
+    options: Omit<T, 'from'>,
   ): Promise<SimulationReturn<T['includeMetadata']>>;
```
Could you explain here why we replace the `from` here with `this.from`?

Thunkar replied:
This is an example of a higher order interaction that should override `ContractFunctionInteraction`. A public authwit can only be set from `msg_sender`, so providing a different `from` than the one used to compute the inner hash doesn't make any sense. Will add a comment

Thunkar replied:
There was a comment 🙃 

<img width="734" height="162" alt="Screenshot 2025-09-30 at 19 48 37" src="https://github.com/user-attachments/assets/2710928f-f95a-404a-a871-c7b3190ef77d" />

benesjan commented on yarn-project/aztec.js/src/wallet/account_manager.ts:
```diff
@@ -13,36 +11,16 @@ import { AccountWithSecretKey } from '../account/account_with_secret_key.js';
     private wallet: Wallet,
-    // TODO: Drop pxe from here once the wallet interface is be updated (then we will not need the direct access to
-    // pxe). Hence here having the type be `any` is not a long-term problem.
-    private pxe: any,
```
🥳

nventuro commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -68,30 +69,41 @@ export class EmbeddedWallet extends BaseWallet {
-   * (who is paying the fee) is unused
-   * @param _address - Unused
+   * unless otherwise stated
+   * @param address - The account building the transaction, usually the one paying the fee
```
There can be txs with no 'account building the tx', eg a multicall that calls into two account contracts. It sounds like this is more about the fee payer than about 'the account building the tx'?

Thunkar replied:
Addressed

nventuro commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -68,30 +69,41 @@ export class EmbeddedWallet extends BaseWallet {
-    const paymentMethod =
-      userFeeOptions?.paymentMethod ??
-      new SponsoredFeePaymentMethod(sponsoredFPCContract.instance.address);
+    let paymentMethod;
+    let endSetup = false;
+    let isFeePayer = false;
+    if (!userFeeOptions?.embeddedPaymentMethodFeePayer) {
+      paymentMethod = new SponsoredFeePaymentMethod(
+        sponsoredFPCContract.instance.address
+      );
+    } else {
+      isFeePayer = address.equals(userFeeOptions.embeddedPaymentMethodFeePayer);
+    }
```
This is a bit convoluted, no? I tried to rewrite it in simpler terms but the logic is all a bit much. It's hard to imagine the different combinations of `isFeePayer`, `paymentMethods`, etc - it seems suspicious that _all_ of them would be valid. If that's the case, maybe we should have proper tagged unions (enums)?

nventuro commented on noir-projects/aztec-nr/aztec/src/authwit/account.nr:
```diff
@@ -27,28 +25,40 @@ impl<Context> AccountActions<Context> {
+    ///
+    /// @param end_setup If paying transactions with fee juice, the account contract itself usually has to signal the end of the setup phase since
+    /// no other contract will do it. This is configurable independently of the previous flag because in the case of using
+    /// FeeJuiceWithClaim the account contract is the fee payer, but the end of the setup is handled by the FeeJuice contract.
```
Remind me why this is the case? Because the FeeJuice claim function is magical and the nodes know about it, and we rely on that fn ending the setup phase?

If so, is this the only instance in which we are the fee payer but we don't end the setup phase? e.g. is `is_fee_payer: false` and `end_setup: true` ever a valid combo?

nventuro replied:
(if it's not, then we might want to instead represent this with enums/unions, e.g. options being:
- someone else pays the fee
- we pay the fee via the fee juice claim
- we pay the fee via some alternative mechanism)

nventuro commented on yarn-project/aztec.js/src/account/index.ts:
```diff
@@ -1,8 +1,5 @@
- * allows to instantiate, initialize and publish a fresh account, or to obtain a `Wallet` instance out of an
- * already-deployed account. Use the `@aztec/accounts` package to load default account implementations that rely
- * on ECDSA or Schnorr signatures.
+ * The `account` module provides utilities for managing accounts.
```
muh docs

Thunkar replied:
These were stale and plainly incorrect

nventuro commented on yarn-project/entrypoints/src/interfaces.ts:
```diff
@@ -19,6 +17,17 @@ export type TxExecutionOptions = {
    */
   txNonce?: Fr;
+  /**
+   * Account contracts usually support paying transaction fees using their own fee juice balance, which is configured
+   * by setting this flag to true.
+   */
+  isFeePayer: boolean;
+  /**
+   * When paying transactions with fee juice, the account contract itself usually has to signal the end of the setup phase since
+   * no other contract will do it. This is configurable independently of the previous flag because in the case of using
+   * FeeJuiceWithClaim the account contract is the fee payer, but the end of the setup is handled by the FeeJuice contract.
+   */
+  endSetup: boolean;
```
This seems to point in the direction of not all combinations being valid, and it this point it looks a bit like an abstraction leak. I'd seriously consider not exposing so much low-level detail and doing e.g. the unions.

nventuro reviewed (approved):
Thanks for going through all of this, it's a bit mind bending how bad it was.

The overall direction is very good and I'm glad this is getting done. I have lots of comments that echo things @benesjan said related to how the different options are handled etc. I think what might be going on is that the thing was _so bad_ you're only now finally understanding which tools are required in order to make sense of it, whereas we just see the new tools and go 'yea these make sense, but couldn't they be nicer?' - which is probably something you didn't even have the time to think about.

So I'm happy about merging things as-is in the interest of avoiding further merge conflicts etc., but I do think we'll want to do another pass on these.

>The FeeJuice contract now has 2 separate claim methods, one that can be invoked anytime, and a specific "fee payment" one that also ends the setup phase.

I think this might be fine. I'd add a comment in `claim_and_end` mentioning that this function is super-special.

Thunkar committed:
feat!: entrypoint and fee payment simplification

# Context for the reviewer

This PR initially attempted to simplify the way fee payment methods were forwarded (if at all!) to the `Wallet` software. Historically, we have allowed these fee payment methods to be assembled "app side" (even if we didn't have a clear separation between wallets and apps). After the wallet refactor and new interface, it became clear this is not  a good design since ultimately wallets (read: users) should decide how the want to pay fees. Unfortunately this is not as simple as that, because:

- Thanks to fee abstraction, fee payment methods are *just contract function calls*. Realistically, there's nothing preventing an app from crafting a transaction that pays for its own fee and sending it to the wallet. We even had a case in our repo: Account contract deployments
- There are legitimate use cases for apps assembling their own fee-paying transactions, such as sponsoring, subscriptions, etc.

With this in mind, now please bear with me

# Entrypoints and fee payment methods

Up until now, entrypoints of account contracts had knowledge of which calls were related to fee payment and which ones were "app" calls. Even if this might seem practical, for the aforementioned reasons it didn't make much sense. Nothing prevented an app from ignoring this convention, and to add insult to injury the account contract was the one responsible of ending the setup (non-revertible) phase of the transaction, which would break apps that tried to put fee calls in the "app" part.

What's even worse, ending the setup of the tx on the account entrypoint lead to the very unfortunate consequence of FPCs never having the assurance of them being executed on that phase. A malicious account contract could just put the calls to pay fees on the revertible side of the tx and make them fail on purpose, while keeping purely private app logic in the non-revertible phase...essentially getting the side effects for free.

# What's changed

- Account contract entrypoints do not distinguish between app and fee calls now. This reflects the fact that these calls are indeed indistinguishable and we should not attempt to add a convention that is broken from the get-go.
- Ending the setup phase is the responsibility of *whomever wants to get paid*. As of now, ending it is the only way contracts can get assurance on whether their side effects are revertible or not.
- Fee payment methods *can* be used on the app side, and we provide some help with that:
    - Interactions (`ContractFunctionInteraction`, `DeployMethod`, etc) allow setting a  `fee.paymentMethod` option just like before. HOWEVER, what this will do is obtain the `ExecutionPayload` for that particular fee payment method and it will merge it with whatever the interaction was doing. This works like this because *it is the only sane way*. If the app wants to "bundle" a fee payment method, it needs intimate knowledge of how the calls are assembled. A great example of this is account contract self-deployments: the fee payment method has to come AFTER the construction of the contract, or it won't be able to validate an authwit (the public key not does not exist yet) and it won't be able to pay for its own fee. This only used to work because we sneakily inverted the order of execution for the `MulticallEntrypoint` to allow for this use case. Terrible.
    - `FeeJuicePaymentMethod` CANNOT be bundled into an interaction. For two reasons: first of all, it's very weird that an app tells the wallet: "hey, here's your tx with a fee payment method, do not a attempt to set one. By the way, the account I'm sending this for is paying". On the other hand, the app would have to convey to the account that they're the fee payer and ALSO that they have to end the setup phase. Conceptually, paying the fee with your own account contract is a *feature* of the account contract itself (which it might disallow for privacy reasons!) so the only one that could possible set it reliably is the wallet.

# Happy consequences

- Account entrypoints can now do 5 functions by default, instead of 4 + to optional fee ones. The entrypoint is also cheaper!
- Fee payment methods can still be bundled on the "app side"
- The `Wallet` interface is now 100% serializable
- FPCs now can be sure that they executed on the setup phase

# Not so happy consequences

- Not being able to say `paymentMethod: FeeJuice` feels weird. Once you think about it, it makes sense (from the app side). The wallet of course will still be able to do that (if its account contracts support it)
- The FeeJuice contract now has 2 separate `claim` methods, one that can be invoked anytime, and a specific "fee payment" one that also ends the setup phase.
- The CLI wallet is in a weird place rn and `cancel_tx` command is broken. Will address it in a followup PR

Co-authored-by: Gregorio Juliana <gregojquiros@gmail.com>
Co-authored-by: thunkar <gregjquiros@gmail.com>
Co-authored-by: thunkar <gregojquiros@gmail.com>
